tools/reviewer/local_server/service/CodeReviewService.java

```diff
@@ -95,78 +97,77 @@ private Repo getHeadRepo(String repoId) {
     return repoFactory.create(repoPath);
   }
 
-  // Returns null when file does not exist in head repo
-  private String readTextFile(File file) throws IOException {
+  private void validateFile(File file) {
     if (file.getFilename().isEmpty()) {
       throw new IllegalArgumentException("File does not have a filename:\n" + file);
     }
+    if (!file.getWorkspace().isEmpty()) {
+      if (!file.getUser().isEmpty()) {
+        // It's another user
+        throw new UnsupportedOperationException("Other users are currently not supported");
+      }
+      if (!file.getCommitId().isEmpty()) {
+        if (!workspaceExists(file.getWorkspace())) {
+          if (!getHeadRepo(file.getRepoId()).commitExists(file.getCommitId())) {
+            logger
+                .atInfo()
+                .log(
+                    "Workspace %s doesn't exist, and commit %s doesn't exist in head repo",
+                    file.getWorkspace(), file.getCommitId());
+            file = file.toBuilder().clearWorkspace().build();
+          }
+        }
+      }
+    }
+  }
+
+  // Fallback to head repo if workspace does not exist
+  private File getHeadFallackIfNeeded(File file) {
+    if (!file.getWorkspace().isEmpty()
+        && !file.getCommitId().isEmpty()
+        && !workspaceExists(file.getWorkspace())) {
+      return file.toBuilder().clearWorkspace().build();
+    }
+    return file;
+  }
+
+  // Returns null when file does not exist
+  private String readHeadFile(File file) throws IOException {
```

Comment:
Exception 'java.io.IOException' is never thrown

Reply:
Done